state-surgery/cli.go

```diff
@@ -0,0 +1,417 @@
+	err = IterateMintEvents(db, headBlock.NumberU64(), func(address common.Address) error {
+		addressesToMigrate[address] = true
+		storageSlotsToMigrate[CalcOVMETHStorageKey(address)] = 1
+		logProgress()
+		return nil
+	})
+	if err != nil {
+		return wrapErr(err, "error reading mint events")
+	}
+
+	// Make sure all addresses are accounted for by iterating over
+	// the OVM ETH contract's state, and panicking if we miss
+	// any storage keys. We also keep track of the total amount of
+	// OVM ETH found, and diff that against the total supply of
+	// OVM ETH specified in the contract.
+	backingStateDB := state.NewDatabase(db)
+	stateDB, err := state.New(root, backingStateDB, nil)
+	if err != nil {
+		return wrapErr(err, "error opening state DB")
+	}
+	storageTrie := stateDB.StorageTrie(OVMETHAddress)
+	storageIt := trie.NewIterator(storageTrie.NodeIterator(nil))
+	logProgress = ProgressLogger(10000, "iterating storage keys")
+	totalFound := new(big.Int)
+	totalSupply := getOVMETHTotalSupply(stateDB)
+	for storageIt.Next() {
+		_, content, _, err := rlp.Split(storageIt.Value)
+		if err != nil {
+			panic(err)
+		}
+
+		k := common.BytesToHash(storageTrie.GetKey(storageIt.Key))
+		v := common.BytesToHash(content)
+		sType := storageSlotsToMigrate[k]
+
+		switch sType {
+		case 1:
+			// This slot is a balance, increment totalFound.
+			totalFound = totalFound.Add(totalFound, v.Big())
+		case 2:
+			// This slot is an allowance, ignore it.
+			continue
+		default:
+			slot := new(big.Int).SetBytes(k.Bytes())
+			// Check if this slot is a variable. If it isn't, and it isn't a
+			// known missing key, abort
+			if slot.Cmp(maxSlot) == 1 && !params.KnownMissingKeys[k] {
+				log.Crit("missed storage key", "k", k.String(), "v", v.String())
+			}
+		}
+
+		logProgress()
+	}
+
+	// Verify that the total supply is what we expect. We allow a hardcoded
+	// delta to be specified in the chain params since older regenesis events
+	// had supply bugs.
+	delta := new(big.Int).Set(totalSupply)
+	delta = delta.Sub(delta, totalFound)
+	if delta.Cmp(params.ExpectedSupplyDelta) != 0 {
+		log.Crit(
+			"supply mismatch",
+			"migrated", totalFound,
+			"supply", totalSupply,
+			"delta", delta,
+			"exp_delta", params.ExpectedSupplyDelta,
+		)
+	}
+
+	log.Info(
+		"supply verified OK",
+		"migrated", totalFound.String(),
+		"supply", totalSupply.String(),
+		"delta", delta.String(),
+		"exp_delta", params.ExpectedSupplyDelta,
+	)
+
+	log.Info("performing migration")
+
+	outDB := MustOpenDB(outDir)
+	outStateDB, err := state.New(common.Hash{}, state.NewDatabase(outDB), nil)
+	if err != nil {
+		return wrapErr(err, "error opening output state DB")
+	}
+
+	// Iterate over the Genesis allocation accounts. These will override
+	// any accounts found in the state.
+	log.Info("importing allocated accounts")
+	logAllocProgress := ProgressLogger(1000, "allocated accounts")
+	for addr, account := range genesis.Alloc {
+		outStateDB.AddBalance(addr, account.Balance)
```

Comment:
Add balance or set balance here? - edit, these are the same b/c balance should always be zero pre-bedrock?

Reply:
Correct, they are identical since balance should be zero pre-Bedrock. There's a check for that above. Swapped it to `SetBalance` anyway since it's clearer.